cmd/zstream/zstream_decompress.c

```diff
@@ -258,32 +263,49 @@ zstream_do_decompress(int argc, char *argv[])
 				default:
 					assert(B_FALSE);
 				}
-				assert(xfunc != NULL);
 
 
 				/*
 				 * Read and decompress the block
 				 */
 				char *lzbuf = safe_calloc(payload_size);
 				(void) sfread(lzbuf, payload_size, stdin);
-				if (0 != xfunc(lzbuf, buf,
+				if (xfunc == NULL) {
+					memcpy(buf, lzbuf, payload_size);
+					drrw->drr_compressiontype =
+					    ZIO_COMPRESS_OFF;
+					if (verbose)
+						fprintf(stderr, "Resetting "
+						    "compression type to off "
+						    "for ino %llu offset "
+						    "%llu\n",
+						    (u_longlong_t)drrw->drr_object,
+						    (u_longlong_t)drrw->drr_offset);
+				} else if (0 != xfunc(lzbuf, buf,
 				    payload_size, payload_size, 0)) {
 					/*
 					 * The block must not be compressed,
-					 * possibly because it gets written
-					 * multiple times in this stream.
+					 * at least not with this compression
+					 * type, possibly because it gets
+					 * written multiple times in this
+					 * stream.
 					 */
 					warnx("decompression failed for "
 					    "ino %llu offset %llu",
 					    (u_longlong_t)drrw->drr_object,
 					    (u_longlong_t)drrw->drr_offset);
 					memcpy(buf, lzbuf, payload_size);
 				} else if (verbose) {
+					drrw->drr_compressiontype =
```

Comment:
Setting `drr_compressiontype` here and on line 307 is just being defensive.  Previously it was actually possible to introduce corruption using `zstream decompress`, if you specified a record that was compressed and had an accurate block pointer.